src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/ordinal/OrdinalRewriterRule.java

```diff
@@ -0,0 +1,210 @@
+import com.alibaba.druid.sql.parser.ParserException;
+import com.alibaba.druid.sql.parser.SQLExprParser;
+import com.alibaba.druid.sql.parser.Token;
+import com.amazon.opendistroforelasticsearch.sql.parser.ElasticSqlExprParser;
+import com.amazon.opendistroforelasticsearch.sql.rewriter.RewriteRule;
+import com.amazon.opendistroforelasticsearch.sql.rewriter.RewriteRuleExecutor;
+import com.amazon.opendistroforelasticsearch.sql.rewriter.alias.TableAliasPrefixRemoveRule;
+import com.amazon.opendistroforelasticsearch.sql.rewriter.identifier.UnquoteIdentifierRule;
+import com.amazon.opendistroforelasticsearch.sql.rewriter.matchtoterm.VerificationException;
+
+import java.sql.SQLFeatureNotSupportedException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Rewrite rule for changing ordinal alias in order by and group by to actual select field.
+ */
+public class OrdinalRewriterRule implements RewriteRule<SQLQueryExpr> {
+
+    private SQLQueryExpr sqlExprGroupCopy;
+    private SQLQueryExpr sqlExprOrderCopy;
+    private final String sql;
+    private final Map<Integer, SQLExpr> groupOrdinalToExpr = new HashMap<>();
+    private final Map<Integer, SQLExpr> orderOrdinalToExpr = new HashMap<>();
+    private RewriteRuleExecutor<SQLQueryExpr> ruleExecutor;
+
+    public OrdinalRewriterRule(String sql) {
+        this.sql = sql;
+    }
+
+    @Override
+    public boolean match(SQLQueryExpr root) throws SQLFeatureNotSupportedException {
+        SQLSelectQuery sqlSelectQuery = root.getSubQuery().getQuery();
+         if (!(sqlSelectQuery instanceof MySqlSelectQueryBlock)) {
+             // it could be SQLUnionQuery
+             return false;
+         }
+
+        MySqlSelectQueryBlock query = (MySqlSelectQueryBlock) sqlSelectQuery;
+
+        if (!hasGroupByWithOrdinals(query) && !hasOrderByWithOrdinals(query)) {
+            return false;
+        }
+
+        // we cannot clone SQLSelectItem, so we need similar objects to assign to GroupBy and OrderBy items
+        sqlExprGroupCopy = generateAST();
+        sqlExprOrderCopy = generateAST();
+
+        collectOrdinalAliases(sqlExprGroupCopy, groupOrdinalToExpr);
+        collectOrdinalAliases(sqlExprOrderCopy, orderOrdinalToExpr);
+
+        if (groupOrdinalToExpr.isEmpty()) {
+            // orderOrdinalToExpr and groupOrdinalToExpr should be identical
+            // because they are generated from similar syntax trees
+            // checking for groupOrdinalToExpr or orderOrdinalToExpr is enough
+            return false;
+        }
+
+        // we need to rewrite the generated syntax tree to match the original syntax tree Select clause
+        ruleExecutor = RewriteRuleExecutor.builder()
+            .withRule(new UnquoteIdentifierRule())
+            .withRule(new TableAliasPrefixRemoveRule())
+            .build();
+
+        ruleExecutor.executeOn(sqlExprGroupCopy);
+        ruleExecutor.executeOn(sqlExprGroupCopy);
```

Comment:
this looks suspicious (the same execution twice)

Reply:
Good catch , the later one should be `sqlExprOrderCopy`.